### ISSUE TITLE
visitCallExpression not traversing through all calls

### ISSUE THREAD

**Issue by cyberhck** [palantir/tslint#4296]
### Bug Report

- __TSLint version__: 5.11.0
- __TypeScript version__: Version 3.1.2
- __Running TSLint via__: CLI

#### TypeScript code being linted

```ts
const something = (cb: () => void) => {};

const selector = something(() => {
  const something = require("typescript");
});
```

with `tslint.json` configuration:

```json
{
  "rules": {
        "no-var-requires": true
  }
}
```

#### Actual behavior
Passes
#### Expected behavior
Fails

But if I change the code for linting to:
```ts
const something = require("typescript");
```
It does fail.

I came across this while working on my own linting rule having a bug, turns out `visitCallExpression` doesn't call for anything inside a callback for some reason.

As of now we've to use normal walker and visit each child of a source file, which might result it in being slow.

I'm not sure if it's limitation of tslint, or typescript not reporting it or something, I'm not even sure if it's known to tslint team.

Please let me know if this is something tslint will be working on? Or is it actually not possible with `visitCallExpression`?

Other rules such as `no-console` actually works because they go through each child.

I did go to AST explorer, and it shows CallExpression.

**Comment by JoshuaKGoldberg** [comment]
Hey @cyberhck - exciting that you're working on your own linting rule, but this post is a little unclear on problem you're facing. Are you saying that `no-var-requires` has a bug because `visitCallExpression` doesn't visit anything, or that the `RuleWalker` class has a bug in `visitCallExpression`?

If it's a bug in `no-var-requires`, that is known and in PR at #2341.

If it's a bug in `RuleWalker`, that would be surprising given how widely used it is. It's generally recommended to use walker functions anyway: see #2522 (just opened #4298 to track marking the walkers as deprecated).

**Comment by JoshuaKGoldberg** [comment]
> it doesn't traverse call expressions

That's probably because you're `return`ing before calling `super.visitCallExpression(node);` in the function. The parent class' `visitCallExpression` is what handles recursively going to the children of that call expression.

> from using `visitCallExpression` to `walk`

Yup! This is preferred. Although `visitCallExpression` seems elegant, it comes with some pretty bad performance issues per that issue.

Search for `applyWithFunction` in the TSLint source code for examples of how to use `walk` style walking. `classNameRule.ts` is a pretty good example of one that only looks for a specific type of node.

Closing this as it's working as intended. Thanks for posting!

**Comment by cyberhck** [comment]
so you're saying, `super.visitCallExpression` is supposed to be called initially (or moving that call to first line will solve it)? I feel stupid, sorry for wasting your time :smile:

**Comment by JoshuaKGoldberg** [comment]
Ha, no worries - this stuff isn't documented very heavily. As long as `super.visitCallExpression` is called, the approach you're taking should work.
